### DISCUSSION TITLE
Proper way to create the main UI when elements take long to be created

### DISCUSSION THREAD

**Discussion by valankar** [zauberzeug/nicegui#3477] [Q&A] [answered]
### Question

I have a page containing many Plotly plots, requiring loading data from disk, etc. They can take a long time to generate.

What's the proper way to do this when it involves the main UI being generated? Most of the examples I've seen involve doing long-running operations after some UI interaction (button click, etc), but in my case the main UI and page load itself can take long to generate.

What I'd generally like to do is have the main page UI show a ui.skeleton/progress, create all the main UI elements in the background, then remove the skeleton/progress and update the UI.

I tried to use background_tasks but can't really figure out the proper way to do it. Is there perhaps a good example I could follow?

Thanks.

**Comment by python-and-novella** [comment]
Maybe running_i_o-bound_tasks https://nicegui.io/documentation/section_action_events#running_i_o-bound_tasks is better.
```python3
import requests
from nicegui import run, ui

async def handle_click():
    URL = 'https://httpbin.org/delay/1'
    response = await run.io_bound(requests.get, URL, timeout=3)
    ui.notify(f'Downloaded {len(response.content)} bytes')

ui.button('Download', on_click=handle_click)

ui.run()
```

**Comment by python-and-novella** [comment]
In nicegui/examples/global_worker/main.py at main · zauberzeug/nicegui
https://github.com/zauberzeug/nicegui/blob/main/examples/global_worker/main.py
```python3
    async def run(self, func: Callable[..., Generator[float, None, None]]) -> None:
        background_tasks.create(run.cpu_bound(self._run_generator, func, self._queue))
        background_tasks.create(self._consume_queue())
```
`background_tasks` is used to create `run.cpu_bound`.

**Comment by falkoschindler** [comment, marked as answer]
Hi @valankar,

You can render some placeholder first and, after the initial page content has been delivered to the client and the socket connection has been established, you can retrieve the actual data and update the page:

```py
import asyncio
from nicegui import ui

async def load_data():
    await asyncio.sleep(1)
    return [
        {'name': 'Alice', 'age': 25},
        {'name': 'Bob', 'age': 30},
        {'name': 'Charlie', 'age': 35},
    ]

@ui.page('/')
async def page():
    with ui.grid(columns=2) as grid:
        for _ in range(3):
            ui.skeleton('text', width='3rem')
            ui.skeleton('text', width='1rem')
    await ui.context.client.connected()
    data = await load_data()
    grid.clear()
    with grid:
        for item in data:
            ui.label(item['name'])
            ui.label(item['age'])

ui.run()
```

**Reply by valankar** [reply]
Thanks! I was able to get it working from your example. I was missing the `await ui.context.client.connected()`. What exactly does that line do? I'm a bit confused by it.

**Reply by falkoschindler** [reply]
Yes, unfortunately this line is rather poorly documented at the moment. Let's break it down:

- `ui.context` holds information about the current UI context. If it is called from a page function, it knows which client has requested this page. (It also holds information about the place in the hierarchy where to put new elements.)
- `client` is the client which requested the page.
- `connected()` is an async function which can be awaited, while the page with its current content is built and send to the frontend. The frontend is instructed to initialize a socket connection. As soon as this connection is created, `connected()` returns and the page function can continue to update the page. This allows the browser to already render an almost complete page while the backend is still doing things like expensive database requests and so on. Everything before `connected()` is delivered as an HTTP response, and everything afterwards is sent in form of update messages over the websocket connection.
